pr_agent/git_providers/github_provider.py

```diff
@@ -222,7 +222,92 @@ def create_inline_comment(self, body: str, relevant_file: str, relevant_line_in_
         return dict(body=body, path=path, position=position) if subject_type == "LINE" else {}
 
     def publish_inline_comments(self, comments: list[dict]):
-        self.pr.create_review(commit=self.last_commit_id, comments=comments)
+        try:
+            self.pr.create_review(commit=self.last_commit_id, comments=comments)
+        except Exception as e:
+            if get_settings().config.verbosity_level >= 2:
+                get_logger().error(f"Failed to publish inline comments, error: {e}")
+            if (
+                getattr(e, "status", None) == 422
+                and get_settings().github.publish_inline_comments_fallback_with_verification
+            ):
+                pass  # continue to try _publish_inline_comments_fallback_with_verification
+            else:
+                raise e
+        try:
+            self._publish_inline_comments_fallback_with_verification(comments)
+        except Exception as e:
+            if get_settings().config.verbosity_level >= 2:
+                get_logger().error(f"Failed to publish inline code comments fallback, error: {e}")
+            raise e
+
+    def _publish_inline_comments_fallback_with_verification(self, comments: list[dict]):
+        """
+        Check each inline comment separately against the GitHub API and discard of invalid comments,
+        then publish all the remaining valid comments in a single review.
+        For invalid comments, also try removing the suggestion part and posting the comment just on the first line.
+        """
+        verified_comments, invalid_comments = self._verify_inline_comments(comments)
+        if invalid_comments and get_settings().github.try_fix_invalid_inline_comments:
+            fixed_comments = self._try_fix_invalid_inline_comments([comment for comment, _ in invalid_comments])
+            verified_fixed_comments, invalid_fixed_comments = self._verify_inline_comments(fixed_comments)
```

Comment:
This line is wrong from my tests. it consistently failed on valid (one-line) code messages

Reply:
A one-line comment isn't necessarily valid, it may be outside the valid range of the hunk (or otherwise invalid). Did your test comment pass when calling `self.pr.create_review` directly?

In my tests whenever I had a comment deemed invalid I then tried to post it directly with `self.pr.create_review` to make sure it indeed failed and it want a false-positive.

Reply:
It deleted for me all valid one-line comments.
and in addition, its extra calls. again

Reply:
About the extra API call: That's true, a fixed message incurs an extra call to the GitHub API, but that makes sense - you create a new message so you have to verify that one as well. It's unfair to compare it to the number of API calls before this PR because there was previously no attempt to fix the invalid message at all - it was just discarded. Also, this is why this feature is behind a separate configuration toggle so you can choose to disable it irrespectively of the primarily feature of this PR (the review comment verification)

Reply:
![image](https://github.com/Codium-ai/pr-agent/assets/21198860/b672b425-7538-420f-8f3b-1fb27ffbab72)